### ISSUE TITLE
Incremental Render Updates

### ISSUE THREAD

**Issue by jmaguirrei** [remotion-dev/remotion#1731]
Hello @JonnyBurger I am wondering about if it's possible to have incremental rendering in the future in Remotion.

Let's say we render a 5 min video and, after that, we realize that some text was wrong. Or we need to change the outro.
In general, let's say that we need to change 5 secs of 5 min video, why do we need to render, again, 4:55 seconds that we (hopefully) know that haven't change?

I imagine this should be possible by having the possibility to persist some "blueprint" of the video, some data structure (returned by the renderer process itself) that represents what was rendered (urls, timestamps, texts, styles, etc...)

Then, if I can feed that snapshot, and the video output url, into a new incremental execution, maybe that execution can make first the difference between the given snapshot and the future one, and act only in the frames that actually changed (this is why we need to provide the video output also).

I know it sounds tricky and may have some limitations (how can we know the data structure matches de video? some token maybe?) but if this is possible, it should be a killer.

My concern is not so much the time and resources spent to get a first render, but the need to iterate over one video several times does.

Let me know your thoughts!

Thanks

**Comment by JonnyBurger** [comment]
It is a good idea and I am glad you bring it up!

I have some concerns about the reliable detection whether a video has changed.
Of course, we could diff the React markup, but there are also external differences like CSS and APIs.

Editing an MP4 video in place is also not straightforward, since an MP4 file works with keyframes, so some frames are interpolated inbetween and do not exist for real, so there will be at least a quality loss.

I don't want to shut down the feature completely yet, but these problems need to be solved in a reliable way since when relying on heuristics, normally something goes wrong 😄

**Comment by jmaguirrei** [comment]
Ok! And thanks for taking the time to answer.

I think there must be limitations.
You mentioned CSS and external APIs, this means they should be cached or avoided at all? Or passed as props?

I see now MP4 is not ideal because interpolation...
Maybe we can go one step earlier in the render process and reuse that as input?
Or maybe we can generate a collection of JPGs (I think s3 can store them if we use lambda) and re-generate only the ones needed?

Let's keep talking, thanks!

**Comment by JonnyBurger** [comment]
@jmaguirrei Yes we could keep the images and re-encode the video. But currently we delete them to save space.

I think CSS and external APIs are useful though! This feature would really also change the model in other ways if we wanted to do it.

I don't have anymore to say at the moment, since I think there is a lack of clean ways to achieve this feature. The Web Platform is really a whole operating system with tons of different variables and serializing this into a state and comparing snapshots is very hard. Proposing to close it as wontfix unless there is a concrete way forward!

**Comment by JonnyBurger** [comment]
Sorry, we are not going to implement this feature.

- This would require keeping the un-encoded frames around, using a lot of disk space. Most people don't want this by default.
- It is hard to determine whether frames have changed. Automatically it cannot be done reliably, and letting you do it by hand will create a very hard to use API.
- It is just not a very popular feature request that we can consider prioritizing.
